File: backend/app/models/history.py
from __future__ import annotations

from sqlalchemy import DateTime, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.models.mixins import IdMixin


class History(IdMixin, Base):
    __tablename__ = "histories"

    user_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("users.id", ondelete="SET NULL"), index=True)
    project_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("projects.id", ondelete="CASCADE"), index=True)
    analysis_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("analyses.id", ondelete="CASCADE"), index=True)
    action: Mapped[str] = mapped_column(String(100), index=True)
    target_type: Mapped[str | None] = mapped_column(String(100))
    target_id: Mapped[str | None] = mapped_column(String(36))
    before_json: Mapped[str | None] = mapped_column(Text)
    after_json: Mapped[str | None] = mapped_column(Text)
    description: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[object] = mapped_column(DateTime, index=True)
    metadata_json: Mapped[str | None] = mapped_column(Text)

File: frontend/src/api/history.ts
import { apiRequest } from "./client";

export interface HistoryRecord {
  id: string;
  user_id?: string | null;
  project_id?: string | null;
  analysis_id?: string | null;
  action: string;
  target_type?: string | null;
  target_id?: string | null;
  description?: string | null;
  created_at: string;
}

export function listHistoryRecords() {
  return apiRequest<HistoryRecord[]>("/api/v1/history");
}
